_posts/2021-01-05-git-rebase-merge.md

```diff
@@ -0,0 +1,167 @@
+
+`git merge` is used for combining the multiple sequences of commits into one unified history. In the most common cases, we use `git merge` to combine two branches. 
+
+Let's take an example in which we will mainly focus on branch merging patterns. In the scenario which we have taken, `git merge` takes two commit pointers, and tries to find the common base commit between them. 
+
+Once Git has found a common base commit, it will create a new "merge commit", that will combine the changes of each queued merge commit sequence.
+
+![Git Merge combining branches](/assets/img/posts/git-rebase-merge/git-merge-working1.png)
+
+![Git Merge with Common Base](/assets/img/posts/git-rebase-merge/git-merge-working2.png)
+
+
+## What is Git Rebase?
+
+Let's have a look at the concept of Git `rebase`. A `rebase` is the way of migrating or combining a sequence of commits to a new base commit. If we consider it in the context of feature branching workflow, then it is more useful and easily visualized. The figure given below describes the general process of rebasing with Git:
+
+![Git Rebasing](/assets/img/posts/git-rebase-merge/git-rebasing-basic.png)
+
+Let's understand the working of Git `rebase` by looking a history with a topic branch off another topic branch. For example, we have branched a topic branch (feature2) from the mainline , and added some feature2-side functionality to our project, and then made a commit. Now, we branch off the `feature2` branch to make some client-side changes. Finally, we go back to the feature2-side branch and commit a few more changes:
+
+![Git Rebase Example](/assets/img/posts/git-rebase-merge/git-rebase-working1.png)
+
+Now suppose that we have decided to merge the client-side changes to the mainline for the release, but we also want to hold the server-side changes until they are tested further. 
+
+With Git `rebase`, we can "replay" the changes in the `feature1` branch (that are not in the `feature2` branch, i.e. C8 and C9), and then replay them on the main branch by using the –onto option of git rebase, We have to specify all the three branches names in this case because we are holding the changes from feature2 branch while replaying them in the main branch from feature1 branch:
+
+```
+git rebase --onto main feature2 feature1
+```
+
+It gives us a bit complex but a pretty cool result:
+
+![Git Rebase Example](/assets/img/posts/git-rebase-merge/git-rebase-working2.png)
+
+
+Now it is time to fast forward our main branch, Remeber that fast forward is simply a unique instance of git rebase when there are no commits to be replayed (and the target branch has new commits while the history of the target branch has not been changed and all the commits on the target branch have the current one as their predecessor.)
+We will use the following commands to do this,
+
+```
+git checkout main
+
+git merge feature1
```

Comment:
From the image I understand that we're merging `feature2` onto master, not `feature1`?? Please check all images that they really say what the text is saying.

Reply:
Yes it's a mistake, updated.
thanks